src/action/ActionManager.js

```diff
@@ -0,0 +1,301 @@
+import KeyboardInputSource from "../input/KeyboardInputSource.js";
+
+/**
+ * If new ActionManager.constructor's loadDefaults param is true then these action maps are loaded
+ */
+const defaultActionMaps = [
+  { name: "default-flat", url: "/js/action-input/default/playing-flat-action-map.json" }
+  /*
+  { name: "default-portal", url: "../default/playing-portal-action-map.json" },
+  { name: "default-immersive", url: "../default/playing-immersive-action-map.json" }
+  */
+];
+
+/**
+ *
+ * TODO figure out how to create links/aliases from specific hw inputs to generic inputs
+ */
+export default class ActionManager {
+  constructor(loadDefaults = true) {
+    this.queryInputPath = this.queryInputPath.bind(this);
+    /**
+     * input semantic path (with wildcards) -> action listener {function}
+     * @type {Map<string, function>}
+     */
+    this._actionListeners = new Map();
+
+    /**
+     * semantic path -> {@link InputSource}
+     * @type {Map<string, InputSource>}
+     */
+    this._inputSources = new Map();
+
+    /**
+     * full semantic path -> {@link Filter}
+     * @type {Map<string, Filter>}
+     */
+    this._filters = new Map();
+
+    /**
+     * name -> {@link ActionMap}
+     * @type {Map<string, ActionMap>}
+     */
+    this._actionMaps = new Map();
+
+    /**
+     * zero or more {@link ActionMap}s may be active at any time
+     * @type {Map<string, ActionMap>}
+     */
+    this._activeActionMaps = new Map();
+
+    /**
+     * actionPath -> { action parameters }
+     * @type {Map<string, Object>}
+     */
+    this._activeActionInfos = new Map();
+
+    // Used during updates
+    this._activations = new Map();
+    this._deactivations = new Map();
+
+    if (loadDefaults) this.loadDefaults();
+  }
+
+  /*
+  @param inputPath {string} a full semantic path for an input, like '/input/keyboard/key/32'
+  @return [inputValue, inputSource]
+  */
+  queryInputPath(inputPath) {
+    for (let inputSourceInfo of this.inputSources) {
+      if (inputPath.startsWith(inputSourceInfo[0]) === false) {
+        continue;
+      }
+      return [inputSourceInfo[1].queryInputPath(inputPath.substring(inputSourceInfo[0].length)), inputSourceInfo[1]];
+    }
+    return [null, null];
+  }
+
+  /**
+   * Poll the input states and update Action state.
+   * If you're rendering inside a requestAnimationFrame response, call this at the top of your render loop.
+   */
+  poll() {
+    // Poll inputs
+    this._inputSources.forEach((inputSource, inputPath) => {
+      inputSource.poll();
+    });
+
+    this._activations.clear();
+    this._deactivations.clear();
+
+    this._activeActionMaps.forEach((actionMap, name) => {
+      actionMap.update(this.queryInputPath, (actionPath, active, actionParameters, inputSource) => {
+        if (active) {
+          // Accept the first activation
+          if (this._activations.has(actionPath)) return;
+          this._activations.set(actionPath, [actionPath, actionParameters, inputSource]);
```

Comment:
Again in this function I see many times used `[actionPath, actionParameters, inputSource]` could we have these stored in a map and reference them instead of creating the new array every time?